File: tests/test_routing_history_store.py
"""Tests for routing/history_store.py — tier-aware history indexing."""

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "src"))

import unittest
from harness_core.dispatch.routing.history_store import RoutingHistoryStore, _task_group_from_row
from harness_core.usage_ledger import UsageLedgerRow


def _make_row(
    profile_id: str = "profile-1",
    cost_group: str = "suite/code/review/quality",
    cost: float = 0.01,
    passed: bool = True,
    input_tokens: int = 100,
    output_tokens: int = 50,
) -> UsageLedgerRow:
    return UsageLedgerRow(
        run_id="run-1",
        case_id="case-1",
        input_tokens=input_tokens,
        output_tokens=output_tokens,
        cached_tokens=0,
        request_count=1,
        tool_call_count=0,
        retry_count=0,
        wall_clock_ms=100,
        estimated_cost=cost,
        pass_=passed,
        cost_of_pass_group=cost_group,
        model_profile_id=profile_id,
        context_pack_id="",
    )


class TaskGroupExtractionTests(unittest.TestCase):
    def test_valid_group(self):
        row = _make_row(cost_group="suite/code/review/quality")
        tg = _task_group_from_row(row)
        self.assertEqual(tg, "code_review")

    def test_invalid_group(self):
        row = _make_row(cost_group="invalid")
        tg = _task_group_from_row(row)
        self.assertIsNone(tg)


class RoutingHistoryStoreTests(unittest.TestCase):
    def test_empty_store(self):
        store = RoutingHistoryStore()
        self.assertEqual(store.total_rows(), 0)
        self.assertIsNone(store.aggregate_by_tier("cheap_executor"))
        self.assertEqual(store.sample_count("code_review"), 0)
        self.assertEqual(store.tiers_observed("code_review"), ())

    def test_add_row(self):
        store = RoutingHistoryStore()
        store.add_row(_make_row(profile_id="p1"))
        self.assertEqual(store.total_rows(), 1)

    def test_tier_profile_map(self):
        store = RoutingHistoryStore(tier_profile_map={"p1": "cheap_executor"})
        self.assertEqual(store.tier_for_profile("p1"), "cheap_executor")
        self.assertIsNone(store.tier_for_profile("p2"))

    def test_set_tier_map(self):
        store = RoutingHistoryStore()
        store.set_tier_map({"p1": "balanced_worker"})
        self.assertEqual(store.tier_for_profile("p1"), "balanced_worker")

    def test_rows_by_tier(self):
        store = RoutingHistoryStore(tier_profile_map={"p1": "cheap_executor", "p2": "balanced_worker"})
        store.add_row(_make_row(profile_id="p1"))
        store.add_row(_make_row(profile_id="p2"))
        store.add_row(_make_row(profile_id="p1"))
        cheap = store.rows_by_tier("cheap_executor")
        balanced = store.rows_by_tier("balanced_worker")
        self.assertEqual(len(cheap), 2)
        self.assertEqual(len(balanced), 1)

    def test_rows_by_task_group(self):
        store = RoutingHistoryStore()
        store.add_row(_make_row(cost_group="suite/code/review/quality"))
        store.add_row(_make_row(cost_group="suite/docs/summarize/speed"))
        store.add_row(_make_row(cost_group="suite/code/review/quality"))
        rows = store.rows_by_task_group("code_review")
        self.assertEqual(len(rows), 2)

    def test_rows_by_tier_and_task_group(self):
        store = RoutingHistoryStore(tier_profile_map={"p1": "cheap_executor", "p2": "balanced_worker"})
        store.add_row(_make_row(profile_id="p1", cost_group="s/code/review/q"))
        store.add_row(_make_row(profile_id="p2", cost_group="s/code/review/q"))
        store.add_row(_make_row(profile_id="p1", cost_group="s/docs/sum/s"))
        rows = store.rows_by_tier_and_task_group("cheap_executor", "code_review")
        self.assertEqual(len(rows), 1)
        self.assertEqual(rows[0].model_profile_id, "p1")

    def test_aggregate_by_tier(self):
        store = RoutingHistoryStore(tier_profile_map={"p1": "cheap_executor"})
        store.add_row(_make_row(profile_id="p1", cost_group="s/c/r/q", cost=0.01, passed=True))
        store.add_row(_make_row(profile_id="p1", cost_group="s/c/r/q", cost=0.02, passed=False))
        agg = store.aggregate_by_tier("cheap_executor")
        self.assertIsNotNone(agg)
        self.assertEqual(agg.total_count, 2)
        self.assertEqual(agg.success_count, 1)
        self.assertEqual(agg.failure_count, 1)

    def test_aggregate_by_tier_and_task_group(self):
        store = RoutingHistoryStore(tier_profile_map={"p1": "cheap_executor"})
        store.add_row(_make_row(profile_id="p1", cost_group="s/code/review/q", cost=0.01, passed=True))
        store.add_row(_make_row(profile_id="p1", cost_group="s/docs/sum/s", cost=0.02, passed=True))
        agg = store.aggregate_by_tier_and_task_group("cheap_executor", "code_review")
        self.assertIsNotNone(agg)
        self.assertEqual(agg.total_count, 1)

    def test_sample_count(self):
        store = RoutingHistoryStore()
        store.add_row(_make_row(cost_group="s/c/r/q"))
        store.add_row(_make_row(cost_group="s/c/r/q"))
        store.add_row(_make_row(cost_group="s/d/s/q"))
        self.assertEqual(store.sample_count("c_r"), 2)
        self.assertEqual(store.sample_count("d_s"), 1)

    def test_sample_count_for_tier(self):
        store = RoutingHistoryStore(tier_profile_map={"p1": "cheap_executor", "p2": "balanced_worker"})
        store.add_row(_make_row(profile_id="p1", cost_group="s/c/r/q"))
        store.add_row(_make_row(profile_id="p2", cost_group="s/c/r/q"))
        self.assertEqual(store.sample_count_for_tier("c_r", "cheap_executor"), 1)

    def test_tiers_observed(self):
        store = RoutingHistoryStore(tier_profile_map={"p1": "cheap_executor", "p2": "balanced_worker"})
        store.add_row(_make_row(profile_id="p1", cost_group="s/c/r/q"))
        store.add_row(_make_row(profile_id="p2", cost_group="s/c/r/q"))
        tiers = store.tiers_observed("c_r")
        self.assertEqual(tiers, ("balanced_worker", "cheap_executor"))

    def test_all_rows(self):
        store = RoutingHistoryStore()
        store.add_row(_make_row())
        store.add_row(_make_row())
        self.assertEqual(len(store.all_rows()), 2)


if __name__ == "__main__":
    unittest.main()

File: engine/tests/test_routing_history_store.rs
use std::collections::HashMap;

use engine::routing::history_store::RoutingHistoryStore;
use engine::routing::schemas::*;

fn make_row(
    id: &str,
    profile_id: &str,
    group: &str,
    cost: f64,
    quality: f64,
    success: bool,
) -> UsageLedgerRow {
    UsageLedgerRow {
        row_id: id.to_string(),
        dispatch_id: format!("disp-{id}"),
        model_profile_id: profile_id.to_string(),
        cost_of_pass_group: group.to_string(),
        input_tokens: 100,
        output_tokens: 50,
        estimated_cost: cost,
        quality_score: quality,
        success,
        failure_domain: None,
        latency_ms: 50,
    }
}

fn make_store() -> RoutingHistoryStore {
    let mut tier_map = HashMap::new();
    tier_map.insert("profile-cheap".to_string(), "cheap_executor".to_string());
    tier_map.insert(
        "profile-balanced".to_string(),
        "balanced_worker".to_string(),
    );
    let mut store = RoutingHistoryStore::new(Some(tier_map));
    store.add_row(make_row(
        "r1",
        "profile-cheap",
        "scope/code/review/accuracy",
        0.01,
        0.8,
        true,
    ));
    store.add_row(make_row(
        "r2",
        "profile-cheap",
        "scope/code/review/accuracy",
        0.02,
        0.6,
        false,
    ));
    store.add_row(make_row(
        "r3",
        "profile-balanced",
        "scope/code/review/accuracy",
        0.03,
        0.9,
        true,
    ));
    store.add_row(make_row(
        "r4",
        "profile-cheap",
        "scope/docs/generate/completeness",
        0.005,
        0.7,
        true,
    ));
    store
}

#[test]
fn test_rows_by_tier() {
    let mut store = make_store();
    let cheap_rows = store.rows_by_tier("cheap_executor");
    assert_eq!(cheap_rows.len(), 3);
    let balanced_rows = store.rows_by_tier("balanced_worker");
    assert_eq!(balanced_rows.len(), 1);
}

#[test]
fn test_rows_by_task_group() {
    let mut store = make_store();
    let review_rows = store.rows_by_task_group("code/review");
    assert_eq!(review_rows.len(), 3);
    let generate_rows = store.rows_by_task_group("docs/generate");
    assert_eq!(generate_rows.len(), 1);
}

#[test]
fn test_rows_by_tier_and_task_group() {
    let mut store = make_store();
    let rows = store.rows_by_tier_and_task_group("cheap_executor", "code/review");
    assert_eq!(rows.len(), 2);
}

#[test]
fn test_tiers_observed() {
    let mut store = make_store();
    let tiers = store.tiers_observed("code/review");
    assert_eq!(tiers.len(), 2);
    assert!(tiers.contains(&"cheap_executor".to_string()));
    assert!(tiers.contains(&"balanced_worker".to_string()));
}

#[test]
fn test_sample_count() {
    let mut store = make_store();
    assert_eq!(store.sample_count("code/review"), 3);
    assert_eq!(store.sample_count("docs/generate"), 1);
    assert_eq!(store.sample_count("nonexistent"), 0);
}

#[test]
fn test_aggregate_by_tier_and_task_group() {
    let mut store = make_store();
    let agg = store
        .aggregate_by_tier_and_task_group("cheap_executor", "code/review")
        .unwrap();
    assert_eq!(agg.total_count, 2);
    assert_eq!(agg.failure_count, 1);
    assert!((agg.total_cost - 0.03).abs() < 0.0001);
}

#[test]
fn test_total_rows() {
    let store = make_store();
    assert_eq!(store.total_rows(), 4);
}

#[test]
fn test_empty_store() {
    let mut store = RoutingHistoryStore::new(None);
    assert_eq!(store.total_rows(), 0);
    assert!(store.tiers_observed("code/review").is_empty());
    assert!(store.aggregate_by_tier("cheap_executor").is_none());
}
